core/include/cubos/core/ecs/event_pipe.hpp

```diff
@@ -0,0 +1,126 @@
+        /// @param event Event which will be inserted into event pipe.
+        /// @param mask Event mask.
+        void push(T event, unsigned int mask = DEFAULT_PUSH_MASK);
+
+        /// Returns the event mask from event pipe at index.
+        /// @param index Event index.
+        unsigned int getEventMask(std::size_t index) const;
+
+        /// Returns the event from event pipe at index (with read/write permissions) and its mask.
+        /// @param index Event index.
+        std::pair<const T&, unsigned int> get(std::size_t index) const;
+
+        /// Clears pipe event list.
+        /// Only events that got read by all readers are deleted!
+        void clear();
+
+        /// Returns pipe event list size.
+        std::size_t size() const;
+
+        /// Adds a new reader to reader count.
+        void addReader();
+
+        /// Removes a reader from reader count.
+        void removeReader();
+
+    private:
+        /// Represents an Event, with its custom type, its mask and its read count.
+        struct Event
+        {
+            T event;
+            unsigned int mask;
+            mutable std::atomic_size_t readCount{0};
+
+            Event(T e, unsigned int m) : event(e), mask(m)
+            {
+            }
+            Event(const Event& other)
+            {
+                *this = other;
+            }
+            Event& operator=(const Event& other)
+            {
+                this->event = other.event;
+                this->mask = other.mask;
+                this->readCount = other.readCount.load();
+                return *this;
+            }
+        };
+
+        /// List of events that are in the pipe.
+        std::deque<Event> events;
+
+        /// Keeps track of how many readers the event pipe currently has.
+        std::size_t readerCount;
+
+        /// Keeps track of how many events were deleted.
+        std::size_t deletedEvents{0};
+    };
+
+    // EventPipe implementation.
+
+    template <typename T> void EventPipe<T>::push(T event, unsigned int mask)
+    {
+        this->events.push_back(Event(event, mask));
+    }
+
+    template <typename T> unsigned int EventPipe<T>::getEventMask(std::size_t index) const
+    {
+        index = index - deletedEvents;
+        return this->events.at(index).mask;
+    }
+
+    template <typename T> std::pair<const T&, unsigned int> EventPipe<T>::get(std::size_t index) const
+    {
+        index = index - deletedEvents;
+        const Event& ev = this->events.at(index);
+        ev.readCount++;
+        return std::pair<const T&, unsigned int>(ev.event, ev.mask);
+    }
+
+    template <typename T> void EventPipe<T>::clear()
+    {
+        while (!this->events.empty() && this->events.front().readCount == this->readerCount)
+        {
+            this->events.pop_front();
+            this->deletedEvents++;
+        }
+    }
+
+    template <typename T> std::size_t EventPipe<T>::size() const
+    {
+        return this->events.size();
```

Comment:
You also need to add `this->deletedEvents` here, since on EventReader it is used to determine if all events have been read.